src/function/table/arrow.cpp

```diff
@@ -275,15 +252,77 @@ static void ArrowScanFunction(ClientContext &context, const FunctionData *bind_d
 			throw std::runtime_error("Unsupported type " + output.data[col_idx].GetType().ToString());
 		}
 	}
+}
+
+static void ArrowScanFunction(ClientContext &context, const FunctionData *bind_data,
+                              FunctionOperatorData *operator_state, DataChunk *input, DataChunk &output) {
+	auto &data = (ArrowScanFunctionData &)*bind_data;
+	auto &state = (ArrowScanState &)*operator_state;
+	//! have we run out of data on the current chunk? move to next one
+	if (state.chunk_offset >= (idx_t)state.chunk->length) {
+		state.chunk_offset = 0;
+		state.ReleaseArray();
+		delete state.chunk;
+		if (data.is_consumed) {
```

Comment:
The bind_data should be read-only, because the same bind_data might be used multiple times when e.g. recursive CTEs are specified (hence why the bind_data is passed in as a constant). The control flow here should be done using `ArrowScanParallelStateNext` and/or happening through the operator state.